File: lib/cosmos/entities/asteroid.rb
# Cosmos Asteroid definition
#
# Copyright (C) 2012-2014 Mohammed Morsi <mo@morsi.org>
# Licensed under the AGPLv3+ http://www.gnu.org/licenses/agpl.txt

require 'cosmos/entity'

module Cosmos
module Entities

# http://en.wikipedia.org/wiki/Asteroid
#
# Cosmos entity residing in a solar system, may be associated with
# resources through {Cosmos::ResourceSource}. Primarily interacted with
# by {Manufactured::Ship} to mine the contained resources.
class Asteroid
  include Cosmos::SystemEntity

  attr_accessor :resources

  CHILD_TYPES = []

  # Cosmos::Asteroid intializer
  def initialize(args = {})
    init_entity(args)
    init_system_entity(args)

    attr_from_args args, :resources => []
  end

  # Return boolean indicating if this asteroid is valid.
  def valid?
    entity_valid? && system_entity_valid? && resources_valid?
  end

  # Return bool indiciating if asteroid location is valid
  def location_valid?
    super && @location.movement_strategy.is_a?(Motel::MovementStrategies::Stopped)
  end

  # return bool inidicating if resources are valid
  def resources_valid?
    @resources.all? { |r| r.valid? }
  end

  # Size doesn't currently apply to asteroid, always validate
  def size_valid?
    true
  end

  # Return boolean indicating if this asteroid can accept the specified resource.
  #
  # TODO right now indiscremenantly accepts all valid resources, make this more selective
  def accepts_resource?(res)
    res.valid?
  end

  # Set resource locally
  def set_resource(res)
    r = @resources.find { |r| r.material_id == res.material_id }
    if r
      # simply update quantity
      if res.quantity > 0
        r.quantity = res.quantity

      # delete resource
      else
        @resources.delete(r)
      end

      return r
    end

    # add resource
    res.entity = self
    @resources << res
    return res
  end

  # Return json representation of asteroid
  def to_json(*a)
    { :json_class => self.class.name,
      :data       => {:resources => @resources }.
                             merge(entity_json).
                      merge(system_entity_json)
    }.to_json(*a)
  end

   # Create new asteroid from json representation
   def self.json_create(o)
     a = new(o['data'])
     return a
   end
end # class Asteroid
end # module Entities
end # module Cosmos

File: site/source/javascripts/omega/asteroid.js
/* Omega Asteroid JS Representation
 *
 * Copyright (C) 2013 Mohammed Morsi <mo@morsi.org>
 *  Licensed under the AGPLv3 http://www.gnu.org/licenses/agpl.txt
 */

//= require "omega/asteroid/commands"
//= require "omega/asteroid/gfx"

Omega.Asteroid = function(parameters){
  this.components = [];
  this.shader_components = [];
  $.extend(this, parameters);

  this.location = Omega.convert.entity(this.location)
};

Omega.Asteroid.prototype = {
  constructor: Omega.Asteroid,
  json_class : 'Cosmos::Entities::Asteroid',

  /// Return bool indicating if asteroid has the specified resource
  has_resource : function(id){
    return !!(this.resource(id));
  },

  /// Return resource for the specified id
  resource : function(id){
    return $.grep(this.resources, function(r){ return r.id == id; })[0];
  },

  /// Return Asteroid in JSON format
  toJSON : function(){
    return {json_class : this.json_class,
            id         : this.id,
            name       : this.name,
            location   : this.location ? this.location.toJSON() : null,
            parent_id  : this.parent_id,
            color      : this.color,
            size       : this.size};
  },

  clicked_in : function(canvas){
    canvas.follow_entity(this);
  }
};

$.extend(Omega.Asteroid.prototype, Omega.AsteroidCommands);
$.extend(Omega.Asteroid.prototype, Omega.AsteroidGfx);
